File: session.hs
{-# LANGUAGE TypeFamilies #-}

data Stop = Done
newtype In  a b = In (a -> IO b)
data    Out a b = Out a   (IO b)

class Session a where
  type Dual a
  run :: a -> Dual a -> IO ()

instance (Session b) => Session (In a b) where
     type Dual (In a b) = Out a (Dual b)
     run (In f) (Out a d) = f a >>= \b -> d >>= \c -> run b c

instance (Session b) => Session (Out a b) where
     type Dual (Out a b) = In a (Dual b)
     run (Out a d) (In f) = f a >>= \b -> d >>= \c -> run c b

instance Session Stop where
     type Dual Stop = Stop
     run  Done Done = return ()

add_server :: In Int (In Int (Out Int Stop))
add_server =  In $ \x -> return $ In $ \y ->
              do { putStrLn "Thinking"
                 ; return $ Out (x + y) (return Done) }

add_client :: Out Int (Out Int (In Int Stop))
add_client =  Out 3 $ return $ Out 4 $
              do { putStrLn "Waiting"
                 ; return $ In $ \z -> print z >> return Done }                

                                       

File: session.scala
object session {

  case class Stop()
  case class In[-A,B](recv: A => B)
  case class Out[+A,B](data: A, cont: B)

  trait Session[S] {
    type Self = S
    type Dual
    type DualOf[D] = Session[Self] { type Dual = D }
    def run(self: Self, dual: Dual): Unit
  }

  implicit object StopDual extends Session[Stop] {
    type Dual = Stop
    def run(self: Self, dual: Dual) = {}
  } 

  implicit def InDual[D,C](implicit cont: Session[C]) = new Session[In[D,C]] {
    type Dual = Out[D,cont.Dual]
    def run(self: Self, dual: Dual) = cont.run(self.recv(dual.data), dual.cont)
  }

  implicit def OutDual[D,C](implicit cont: Session[C]) = new Session[Out[D,C]] {
    type Dual = In[D,cont.Dual]
    def run(self: Self, dual: Dual) = cont.run(self.cont, dual.recv(self.data))
  }

  def run[S,D:Session[S]#DualOf](session: S, dual: D) = {
    implicitly[Session[S]#DualOf[D]].run(session, dual) 
  }

  def add_server = {
    In { x: Int => {
      In { y: Int => {
        println("Thinking")
        Out (x + y, Stop())
      }}
    }}
  }

  def add_client = {
    Out ( 3, {
      Out ( 4, {
        println("Waiting")
        In { z: Int => {
          println(z)
          Stop()
        }}
      })
    })
  }

  def main = run(add_server, add_client)
}
